File: iCyption2022/CftTest/segmen/solution.py
class Point:
    def __init__(self, x = 0, y = 0):
        self.x = x
        self.y = y
    def __eq__(self, p):
        return self.x == p.x and self.y == p.y
    def __add__(self, p):
        return Point(self.x + p.x, self.y + p.y)
    def __sub__(self, p):
        return Point(self.x - p.x, self.y - p.y)
    def __mul__(self, d):
        return Point(self.x*d, self.y*d)
    def dot(self, p):
        return self.x*p.x + self.y*p.y
    def cross(self, p):
        return self.x*p.y - self.y*p.x
    def dist2(self):
        return self.x**2 + self.y**2

n = int(input())

vs = []
exist = {}

def addToExist(p, val):
    if p.x in exist:
        exist[p.x][p.y] = val
    else:
        d = dict()
        d[p.y] = val
        exist[p.x] = d

for i in range(n):
    x1, y1, x2, y2 = map(int, input().split())
    a, b = Point(x1, y1), Point(x2, y2)
    vs.append(b - a)
    addToExist(vs[i], i)

def findThirdSegment(segments):
    # try every possibility of positive/negatives
    for bit in range(4):
        Sum = Point(0, 0)
        for i in range(2):
            if bit & 2**i != 0:
                Sum = Sum + segments[i]
            else: 
                Sum = Sum - segments[i]

        if Sum.x in exist and Sum.y in exist[Sum.x]:
            return exist[Sum.x][Sum.y]
        min1 = Sum*-1
        if min1.x in exist and min1.y in exist[min1.x]:
            return exist[min1.x][min1.y]
    return -1

maxPerim = -1
ans = [0, 0, 0]
for i in range(n):
    for j in range(i + 1, n):
        if vs[i].cross(vs[j]) == 0: continue

        thirdSegment = findThirdSegment([vs[i], vs[j]])
        if thirdSegment != -1:
            perim = vs[i].dist2() + vs[j].dist2() + vs[thirdSegment].dist2()
            if perim > maxPerim:
                maxPerim = perim
                ans = [i + 1, j + 1, thirdSegment + 1]

if maxPerim == -1:
    print(-1)
else:
    print(ans[0], ans[1], ans[2])

File: iCyption2022/CftTest/segmen/solution.cpp
#include <bits/stdc++.h>
using namespace std;

// O(N^2) solution

struct Point {
    typedef Point P;
    int x, y;
    void read() { cin >> x >> y; }
    Point(int _x = 0, int _y = 0) : x(_x), y(_y) {}
    bool operator < (P p) const { return tie(x,y) < tie(p.x,p.y); }
    bool operator == (P p) const { return tie(x,y) == tie(p.x,p.y); }
    P operator + (P& p) { return P(x + p.x, y + p.y); }
    P operator - (P& p) { return P(x - p.x, y - p.y); }
    P operator * (int d) { return P(x*d, y*d); }
    P operator / (int d) { return P(x/d, y/d); }
    int dot(P p) { return x*p.x + y*p.y; }
    int cross(P p) { return x*p.y - y*p.x; }
    int dist2() { return x*x + y*y; }
};

int main() {
    ios::sync_with_stdio(0), cin.tie(0);
    int n;
    cin >> n;

    vector<Point> vs;
    map<Point, int> exist;
    for (int i = 0; i < n; i++) {
        Point a, b;
        a.read(), b.read();
        vs.push_back(b - a);
        exist[vs[i]] = i;
    }

    auto findThirdSegment = [&](array<Point, 2> segments) {
        // try every possibility of positive/negatives
        for (int bit = 0; bit < 4; bit++) {
            Point sum(0, 0);
            for (int i = 0; i < 2; i++) {
                if (bit & (1<<i)) sum = sum + segments[i];
                else sum = sum - segments[i];
            }

            if (exist.find(sum) != exist.end()) return exist[sum];
            if (exist.find(sum*-1) != exist.end()) return exist[sum*-1];
        }
        return -1;
    };

    int maxPerim = -1;
    array<int, 3> ans;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            // to rule out 0-area triangles
            if (vs[i].cross(vs[j]) == 0) continue;

            int thirdSegment = findThirdSegment({vs[i], vs[j]});
            if (thirdSegment != -1) {
                int perim = vs[i].dist2() + vs[j].dist2() + vs[thirdSegment].dist2();
                if (perim > maxPerim) {
                    maxPerim = perim;
                    ans = {i + 1, j + 1, thirdSegment + 1};
                }
            }
        }
    }

    if (maxPerim == -1) {
        cout << -1 << '\n';
    } else {
        cout << ans[0] << ' ' << ans[1] << ' ' << ans[2] << '\n';
    }
}
